rustc 1.92.0-beta3 flags unreachable code where there is none

### Code

```Rust
use std::convert::Infallible;

pub fn foo(f: impl FnOnce () -> Infallible) -> Infallible {
    f()
}
```

### Current output

```Shell
error[E0601]: `main` function not found in crate `lib`
 --> lib.rs:5:2
  |
5 | }
  |  ^ consider adding a `main` function to `lib.rs`

warning: unreachable expression
 --> lib.rs:5:1
  |
4 |     f()
  |     --- any code following this expression is unreachable
5 | }
  | ^ unreachable expression
  |
note: this expression has type `Infallible`, which is uninhabited
 --> lib.rs:4:5
  |
4 |     f()
  |     ^^^
  = note: `#[warn(unreachable_code)]` (part of `#[warn(unused)]`) on by default

error: aborting due to 1 previous error; 1 warning emitted

For more information about this error, try `rustc --explain E0601`.
```

### Desired output

```Shell
None. Compilation should succeed with no warnings.
```

### Rationale and extra context

The real-world version of this where the lint tripped upon updating the compiler looks like this:

```
pub fn reset_stack_guards(continuation: impl FnOnce() -> Infallible) -> Infallible {
    sys::android_reset_stack_guards();
    continuation()
}
```

It is intended that the passed continuation never returns, and no user code exists past it. The fact that it will never return is surfaced to the caller through the type signature.

I understand that there may be MIR which is "unreachable" in this case (the return path), but there is no unreachable user code.

There's currently no way to write the function above without suppressing the modified lint that I can see, because the lint also triggers on

```
pub fn reset_stack_guards(continuation: impl FnOnce() -> Infallible) -> ! {
    sys::android_reset_stack_guards();
    match continuation() {}
}
```

If we had never type support, this lint might be reasonable as with `#![feature(never_type)]` and a signature `fn reset_stack_guards(continuation: impl FnOnce() -> !) -> !`, the lint does not complain.

Since never types are not stable, this means that a useful, default-enabled lint now reports in non-actionable situations.

### Other cases

```Rust

```

### Rust Version

```Shell
rustc 1.92.0-beta.3 (f3f12444a 2025-11-09)
binary: rustc
commit-hash: f3f12444a017add0468f683f3a10656d29438a05
commit-date: 2025-11-09
host: x86_64-unknown-linux-gnu
release: 1.92.0-beta.3
LLVM version: 21.1.3
```

### Anything else?

_No response_

--- a/compiler/rustc_mir_build/src/builder/mod.rs
+++ b/compiler/rustc_mir_build/src/builder/mod.rs
@@ -839,6 +839,25 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
                 self.infcx.typing_env(self.param_env),
             );
 
+            // check if the function's return type is inhabited
+            // this was added here because of this regression
+            // https://github.com/rust-lang/rust/issues/149571
+            let output_is_inhabited =
+                if matches!(self.tcx.def_kind(self.def_id), DefKind::Fn | DefKind::AssocFn) {
+                    self.tcx
+                        .fn_sig(self.def_id)
+                        .instantiate_identity()
+                        .skip_binder()
+                        .output()
+                        .is_inhabited_from(
+                            self.tcx,
+                            self.parent_module,
+                            self.infcx.typing_env(self.param_env),
+                        )
+                } else {
+                    true
+                };
+
             if !ty_is_inhabited {
                 // Unreachable code warnings are already emitted during type checking.
                 // However, during type checking, full type information is being
@@ -849,7 +868,7 @@ impl<'a, 'tcx> Builder<'a, 'tcx> {
                 // uninhabited types (e.g. empty enums). The check above is used so
                 // that we do not emit the same warning twice if the uninhabited type
                 // is indeed `!`.
-                if !ty.is_never() {
+                if !ty.is_never() && output_is_inhabited {
                     lints.push((target_bb, ty, term.source_info.span));
                 }
 

--- a/tests/ui/uninhabited/uninhabited-unreachable-warning-149571.rs
+++ b/tests/ui/uninhabited/uninhabited-unreachable-warning-149571.rs
@@ -1,0 +1,10 @@
+#![deny(unreachable_code)]
+//@ run-pass
+
+use std::convert::Infallible;
+
+pub fn foo(f: impl FnOnce() -> Infallible) -> Infallible {
+    f()
+}
+
+fn main() {}

